rust/lance/src/dataset/optimize.rs

```diff
@@ -85,24 +97,64 @@
 //! you wish. As long as the tasks don't rewrite any of the same fragments,
 //! they can be committed in any order.
 use std::borrow::Cow;
+use std::collections::{HashMap, HashSet};
 use std::ops::{AddAssign, Range};
-use std::sync::Arc;
+use std::sync::{Arc, RwLock};
 
+use arrow_array::{RecordBatch, UInt64Array};
+use async_trait::async_trait;
+use datafusion::error::Result as DFResult;
+use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
 use datafusion::physical_plan::SendableRecordBatchStream;
 use futures::{StreamExt, TryStreamExt};
+use roaring::{RoaringBitmap, RoaringTreemap};
 use serde::{Deserialize, Serialize};
+use uuid::Uuid;
 
+use crate::dataset::ROW_ID;
+use crate::format::RowId;
 use crate::io::commit::commit_transaction;
 use crate::Result;
 use crate::{format::Fragment, Dataset};
 
 use super::fragment::FileFragment;
-use super::transaction::{Operation, RewriteGroup, Transaction};
+use super::index::DatasetIndexRemapperOptions;
+use super::transaction::{Operation, RewriteGroup, RewrittenIndex, Transaction};
 use super::{write_fragments, WriteMode, WriteParams};
 
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct RemappedIndex {
+    original: Uuid,
+    new: Uuid,
+}
+
+impl RemappedIndex {
+    pub fn new(original: Uuid, new: Uuid) -> Self {
+        Self { original, new }
+    }
+}
+
+/// When compaction runs the row ids will change.  This typically means that
+/// indices will need to be remapped.  The details of how this happens are not
+/// a part of the compaction process and so a trait is defined here to allow
+/// for inversion of control.
+#[async_trait]
+pub trait IndexRemapper: Send + Sync {
+    async fn remap_indices(
+        &self,
+        index_map: HashMap<u64, Option<u64>>,
+        fragments: &[Fragment],
+    ) -> Result<Vec<RemappedIndex>>;
+}
+
+/// Options for creating an [IndexRemapper]
+pub trait IndexRemapperOptions: std::fmt::Debug + Default + Clone + PartialEq + Serialize {
+    fn create_remapper(&self, dataset: &Dataset) -> Result<Box<dyn IndexRemapper>>;
+}
+
 /// Options to be passed to [compact_files].
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
-pub struct CompactionOptions {
+pub struct CompactionOptions<I: IndexRemapperOptions = DatasetIndexRemapperOptions> {
```

Comment:
My idea here was that "index remapping" should be completely opaque to compaction.  In practice, there are no options that the remapping process needs and there is only one real implementation of this trait.  We could probably implement this in some much simpler way.  Any opinions?